### ISSUE TITLE
LayersControl is sometimes not correctly updating its layer list, displaying overlays that have been removed

### ISSUE THREAD

**Issue by hughlomas** [PaulLeCam/react-leaflet#175]
`LeafletElement.LayersControl._layers` is mismatching state with the actual `react-leaflet.LayersControl.Overlay` children that are rendered.

I am using `react-leaflet@0.11.6` and `react@15.0.1` and `react-dom@15.0.1`

I have a set of `LayersControl.Overlay` that are being generated based on the current set of markers.

I group my markers into three categories, "active", "paused", and "unknown". 

When updating the marker set to not include any paused or unknown markers, the LayersControl still displays them as options, but unchecked. In other words, one Overlay is rendered in `render()` but the LayersControl still shows phantom entries for the previous Overlays. When the markers are readded, the LayersControl adds new entries for the Overlays of those groups, and leaves the old phantom ones as well. You can see a picture of the results of this behavior below. 

![image](https://cloud.githubusercontent.com/assets/187210/16059189/e713466e-324f-11e6-9c56-8c52fefcc750.png)

![image](https://cloud.githubusercontent.com/assets/187210/16055961/5f30ee80-3241-11e6-8a40-fbdc500c5558.png)

### Expected behavior
The Leaflet control should accurately reflect the current elements of LayersControl.Overlays.

### Actual behavior
LayersControl keeps displaying the name keys of Overlays that were not removed in later `render()` calls, and then ignores them when readding Overlays with the same keys at a later point.

### Steps to reproduce
I have so far been unable to reproduce the issue in a fiddle, despite trying to set up a similar situation.

My "production" code is the following: 

```javascript
render: function () {

    var agents = this.state.agents;

    var agentLayers = _( agents )
      .filter( agent => ( agent.radius && agent.lat && agent.lng ) )
      //returns either "active", "paused", or "unknown"
      .groupBy( this.categorizeAgent )
      .map( ( agents, name ) => {
        var markers = _.map( agents, this.renderAgentMarker );
        return (
          <Overlay checked name={name} key={name}>
            <LayerGroup>
              {markers}
            </LayerGroup>
          </Overlay>
        );
      })
      .valueOf();

    _.each( agentLayers, layer => console.log( layer.key ) );

    return (
      <div id="map-screen">
        <Map
          id="agent-map"
          ref="map"
          center={[Number(this.state.lat), Number(this.state.lng)]}
          zoom={Number(this.state.zoom)}
          animate={true}
          onDragend={this.handleDragEnd}
          onZoomend={this.handleZoomEnd}
          onClick={this.handleMapClick}
        >
          <ScaleControl />
          <LayersControl ref="layersControl" position='topright'>
            <BaseLayer checked name='OpenStreetMap'>
              <TileLayer
                attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
                url='http://{s}.tile.osm.org/{z}/{x}/{y}.png'
              />
            </BaseLayer>
            {agentLayers}
          </LayersControl>
        </Map>
      </div>
    );
  }
```

Also thank you and all contributors for this library, it vastly simplified my marker state management.

**Comment by jgimbel** [comment]
`.groupBy( this.categorizeAgent )`

Are you absolutely sure this is grouping correctly?
From the screenshot and code you have given my first guess would be that you are creating a Overlay for every marker, instead of an overlay for each of your names. Do you have [react-devtools](https://github.com/facebook/react-devtools) installed? From there you could check how many Overlay components are being created, and know if your groupby is working correctly.

**Comment by PaulLeCam** [comment]
Hi,
Indeed it's a bug, thanks for the detailed issue, it helped a lot!
I just published v0.11.7 that should solve the issue, if you can give it a try?

**Comment by hughlomas** [comment]
@jgimbel Yes, the grouping is working correctly as verified with console logs.

@PaulLeCam Thank you for such a quick response, I tried out the fix and it works perfectly!
